plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MavenMetadata.java

```diff
@@ -0,0 +1,314 @@
+    }
+
+    public String getLatest() {
+      return latest;
+    }
+
+    @Nullable
+    public String getRelease() {
+      return release;
+    }
+
+    public List<String> getVersions() {
+      return versions;
+    }
+  }
+
+  // V level
+
+  public static class Snapshot
+  {
+    private final DateTime lastUpdated;
+
+    private final String extension;
+
+    @Nullable
+    private final String classifier;
+
+    private final String version;
+
+    public Snapshot(final DateTime lastUpdated,
+                    final String extension,
+                    @Nullable final String classifier,
+                    final String version)
+    {
+      this.lastUpdated = checkNotNull(lastUpdated, "lastUpdated");
+      this.extension = checkNotNull(extension, "extension");
+      this.classifier = classifier;
+      this.version = checkNotNull(version, "version");
+    }
+
+    public DateTime getLastUpdated() {
+      return lastUpdated;
+    }
+
+    public String getExtension() {
+      return extension;
+    }
+
+    @Nullable
+    public String getClassifier() {
+      return classifier;
+    }
+
+    public String getVersion() {
+      return version;
+    }
+  }
+
+  public static class Snapshots
+  {
+    private final long snapshotTimestamp;
+
+    private final int snapshotBuildNumber;
+
+    private final List<Snapshot> snapshots;
+
+    public Snapshots(final long snapshotTimestamp,
+                     final int snapshotBuildNumber,
+                     final List<Snapshot> snapshots)
+    {
+      this.snapshotTimestamp = snapshotTimestamp;
+      this.snapshotBuildNumber = snapshotBuildNumber;
+      if (snapshots != null) {
+        this.snapshots = ImmutableList.copyOf(snapshots);
+      }
+      else {
+        this.snapshots = null;
+      }
+    }
+
+    public long getSnapshotTimestamp() {
+      return snapshotTimestamp;
+    }
+
+    public int getSnapshotBuildNumber() {
+      return snapshotBuildNumber;
+    }
+
+    public List<Snapshot> getSnapshots() {
+      return snapshots;
+    }
+  }
+
+  // instance
+
+  private final Level level;
+
+  private final DateTime lastUpdated;
+
+  private final String groupId;
+
+  @Nullable
+  private final String artifactId;
+
+  @Nullable
+  private final String version;
+
+  @Nullable
+  private final List<Plugin> plugins;
+
+  @Nullable
+  private final BaseVersions baseVersions;
+
+  @Nullable
+  private final Snapshots snapshots;
+
+  private MavenMetadata(final Level level,
+                        final DateTime lastUpdated,
+                        final String groupId,
+                        @Nullable final String artifactId,
+                        @Nullable final String version,
+                        @Nullable final List<Plugin> plugins,
+                        @Nullable final BaseVersions baseVersions,
+                        @Nullable final Snapshots snapshots)
+  {
+    this.level = checkNotNull(level);
+    this.lastUpdated = lastUpdated;
```

Comment:
The private ctor is used from static helpers only, that does check for NPE. Am on the fence with this...